src/puppetlabs/puppetdb/query_eng.clj

```diff
@@ -82,6 +82,49 @@
                     (:queryable? projection-value))]
      (keyword projection-key))))
 
+(defn legacy-engine-query-context
+  "Parse a query and possibly some paging clauses from a query to the legacy
+   engine. Required because top-level paging options supplied in the query are
+   not otherwise handled by the engine, so they must be merged with
+   param-supplied paging options and dealt with at the top level."
+  [version query warn]
+  (cm/match
+    query
+    ["from" (entity-str :guard #(string? %)) & remaining-query]
+    (let [remaining-query (cm/match
+                            remaining-query
+
+                            [(c :guard eng/paging-clause?) & clauses]
+                            (let [paging-clauses (eng/create-paging-map (cons c clauses))]
+                              {:paging-clauses paging-clauses :query []})
+
+                            [(q :guard vector?)]
+                            {:query q}
+
+                            [(q :guard vector?) & clauses]
+                            (let [paging-clauses (eng/create-paging-map clauses)]
+                              {:query q :paging-clauses paging-clauses})
+
+                            []
+                            {:query []}
+                            :else (throw
+                                    (IllegalArgumentException.
+                                      (str
+                                        (tru "Your `from` query accepts an optional query only as a second argument.")
+                                        " "
+                                        (tru "Check your query and try again.")))))
+          entity (keyword (utils/underscores->dashes entity-str))]
+      (when warn
+        (eng/warn-experimental entity))
+      {:remaining-query (:query remaining-query)
+       :paging-clauses (:paging-clauses remaining-query)})
+
+    :else (throw (IllegalArgumentException.
+                   (str
+                     (trs "Your initial query must be of the form: [\"from\",<entity>,(<optional-query>)].")
```

Comment:
not sure -- to indicate that it's optional you mean? I think the comma inside the parens will look weird but I think we should have spaces after the commas. I'm not sure if the parens are necessary at all (this was just a preexisting message).